### PR TITLE
[ROCm] Re-order torch import to avoid race with rocprofiler

### PR DESCRIPTION
**Pull request by alugorey**
monarch's native extension links the system libamdhip64 while torch bundles its own; when monarch._rust_bindings was imported before torch, the system HIP runtime won rocprofiler-register and   caused `hipApiName has non-null function pointer 1 despite this being the first instance of the library being copies`. In this PR we ensure torch is imported first. The comments above the original lines insinuated this was supposed to already happen, but the code was written in the reverse.

Additionally, we fix a CUdeviceptr bug:
CUdeviceptr is an integer on CUDA but a raw pointer (hipDeviceptr_t) on ROCm, so initialize portably with zeroed (0 / null)

### CODE REVIEW

**Comment by alugorey** [comment]
FYI @samlurye @dulinriley @jithunnair-amd

**Comment by alugorey** [comment]
Hi @samlurye , Could i get you to look at the most recent commit in this PR? the first one fixed the RDMA test failure but it spawned a different failure in test_rdma_buffer_cpu_memoryview which complained about torch being present in sys.modules which broke the contract in the test. the second commit fixed the issue but i want to make sure I don't go down the wrong path here.

**Comment by alugorey** [comment]
@samlurye @dulinriley @mreso Finally got ROCm CI green with this one. Could you please take a look? I did what i had to do to get CI green but i'm sure there are things i didn't take into account or violated some contracts that we'd like to preserve so please look over. We can use this as the jumping off point at least.

File: python/monarch/_src/rdma/rdma.py [hunk 1]
```diff
@@ -221,9 +221,49 @@ def _evict_local_memory(key: Tuple[int, int, int], ref: "weakref.ref") -> None:
...
+        with open("/proc/self/maps") as f:
+            maps = f.read()
+    except OSError:
+        return
+    import re
+
+    libs = set(re.findall(r"/\S*libamdhip64\.so[.0-9]*", maps))
+    if len(libs) < 2:
+        return  # single HIP runtime: safe
+    raise RuntimeError(
```
Comment (dulinriley) [inline python/monarch/_src/rdma/rdma.py]: Maybe print all the library paths found in this case?

**Review by dulinriley — APPROVED** [review]

**Comment by dulinriley** [comment]
@alugorey I had to make a slight patch to fix formatting and a real bug related to device masking. Please update this PR with this patch file:
(use `git am` to apply because this was created with `format-patch`)
[fix-rocm.patch](https://github.com/user-attachments/files/30363198/fix-rocm.patch)

**Comment by alugorey** [comment]
> @alugorey I had to make a slight patch to fix formatting and a real bug related to device masking. Please update this PR with this patch file: (use `git am` to apply because this was created with `format-patch`) [fix-rocm.patch](https://github.com/user-attachments/files/30363198/fix-rocm.patch)

@dulinriley Patch applied

**Comment by samlurye** [comment]
@alugorey Without this fix, does the `hipApiName has non-null function pointer` error happen always on a call from monarch into hip, or can it also happen on a torch api call? In other words, if you first import `monarch._rust_bindings`, then import `torch` and do something with a hip tensor, what happens?

**Review by dulinriley — APPROVED** [review]

File: python/monarch/__init__.py [hunk 2]
```diff
@@ -9,15 +9,67 @@
 from importlib import import_module as _import_module
 from typing import TYPE_CHECKING
 
-# Import before monarch to pre-load torch DSOs as, in exploded wheel flows,
-# our RPATHs won't correctly find them.
+import os as _os
```
Comment (dulinriley) [inline python/monarch/__init__.py]: Are you sure you applied the patch? The internal formatter is complaining about these import positions

**Review by dulinriley — CHANGES_REQUESTED** [review]

File: python/monarch/__init__.py (hunk 2 above)
Comment (dulinriley) [reply python/monarch/__init__.py]: Sorry about this, I wish I could just apply the formatting changes directly, but the internal code and the PR need to be in sync and I'm not allowed to update your branch on your fork. (if you make a branch directly on meta-pytorch/monarch I could edit that directly).

Comment (alugorey) [reply python/monarch/__init__.py]: ah right. so the patch didn't actually seem to apply cleanly. much of the patch had changes that were already in this PR and caused collisions. Claude also said that the patch was a "Mercurial" export which didn't make sense to me. I went ahead and manually added the changes so i may have missed something. Can you re-inspect the patch and make sure it was built right on top of this PR? I don't believe it should have any changes in the diff that are already present in this PR. Also not sure why claude said it was a "mercurial" export or what that even means. in any case, the patch did not apply cleanly using `git am` which is why i pivoted the approach. I also don't think i have the permissions to make a branch directly on the meta-pytorch/monarch repo.

**Comment by alugorey** [comment]
> @alugorey Without this fix, does the `hipApiName has non-null function pointer` error happen always on a call from monarch into hip, or can it also happen on a torch api call? In other words, if you first import `monarch._rust_bindings`, then import `torch` and do something with a hip tensor, what happens?

@samlurye it's less about monarch to hip and torch to hip. the issue happens because there are two copies of libamdhip64 in our case. one on the system (usually in /opt/rocm) and then one bundled with torch. So before this change it became a race to register the hip runtimes with rocprofiler. Whichever runtime registered with rocprofiler **second** would abort. the problem requires both torch and monarch to use hip in the same process (like in the unit tests).

import `monarch._rust_bindings`, then import `torch` and do something with a hip tensor, what happens
this would **not** crash as the system copy is loaded but never initialized. the order dependent symptom would look as follows: 
- monarch-second -> hipApiName SIGABRT (rdmaxcel path)
- torch-second -> torch "No CUDA GPUs are available"

**Comment by alugorey** [comment]
@dulinriley i added you as a collaborator to my fork. perhaps you can re-try adding your changes, if that makes it easier?

**Comment by samlurye** [comment]
> import monarch._rust_bindings, then import torch and do something with a hip tensor, what happens
this would not crash as the system copy is loaded but never initialized. the order dependent symptom would look as follows

@alugorey Instead of relying on import order for correctness (like in this PR), which won't work if someone pulls in a python module that loads hip but that isn't torch, I'm wondering if a more sustainable solution would be to just ensure that monarch never loads the system hip.

Currently, there are two places in monarch's build that link against libamdhip64:
- The `nccl-sys` crate, which belongs to the tensor engine feature and which no one is using.
- The `rdmaxcel-sys` crate, but the symbols requiring libamdhip64 linkage are used only in kernel-launched RDMA functionality, which also currently isn't being used (and may be broken anyway).

So my proposal would be to do the following:
- Decouple the RDMA build from the `USE_TENSOR_ENGINE` flag. Making RDMA usable without the tensor engine will mean that `nccl-sys` won't be built, so that module will never load hip.
- Gate kernel-launched RDMA behind a new flag, default off. Then for common usage, the RDMA build won't need to link against hip and so monarch will never load hip.

What do you think?

File: python/monarch/__init__.py (hunk 2 above)
Comment (dulinriley) [reply python/monarch/__init__.py]: Oh haha mercurial is the git alternative we use inside Meta: https://www.mercurial-scm.org/

That's interesting it rejected the patch because I used git to apply it as a test and it seemed to apply cleanly. But no matter, I can just checkout your branch now and directly apply the formatter.

**Comment by dulinriley** [comment]
You should work with @samlurye to address the comments. Once that's done either he or I can apply any formatting changes needed and re-import

**Comment by alugorey** [comment]
> > import monarch._rust_bindings, then import torch and do something with a hip tensor, what happens
> > this would not crash as the system copy is loaded but never initialized. the order dependent symptom would look as follows
> 
> @alugorey Instead of relying on import order for correctness (like in this PR), which won't work if someone pulls in a python module that loads hip but that isn't torch, I'm wondering if a more sustainable solution would be to just ensure that monarch never loads the system hip.
> 
> Currently, there are two places in monarch's build that link against libamdhip64:
> 
> * The `nccl-sys` crate, which belongs to the tensor engine feature and which no one is using.
> * The `rdmaxcel-sys` crate, but the symbols requiring libamdhip64 linkage are used only in kernel-launched RDMA functionality, which also currently isn't being used (and may be broken anyway).
> 
> So my proposal would be to do the following:
> 
> * Decouple the RDMA build from the `USE_TENSOR_ENGINE` flag. Making RDMA usable without the tensor engine will mean that `nccl-sys` won't be built, so that module will never load hip.
> * Gate kernel-launched RDMA behind a new flag, default off. Then for common usage, the RDMA build won't need to link against hip and so monarch will never load hip.
> 
> What do you think?

I'm totally okay with that! honestly, it was kind of a pain in the butt trying to handle all the ordering edge cases, so if we can sever the need to load libamdhip64 altogether and allow pytorch to just manage that however it wants then that's a much cleaner solution. Give me a day or so to work on that and i'll circle back.
